
<div align="center">

# 📊 Cohort & Retention Analytics Portfolio
### Lộ trình học · 7 Buổi · Python for Data Analysis

</div>

---

## 🗺️ Lộ trình khóa học

| # | Buổi học | Thời lượng | Chủ đề chính | Trạng thái |
|:-:|----------|:----------:|--------------|:----------:|
| 1 | Giới thiệu dự án | 45 phút | Business context · Deliverables · Portfolio standards | ✅ Hoàn thành |
| 2 | Business Brief & Setup Project | 60 phút | Project structure · Dependencies · Business brief · Metric dictionary | ✅ Hoàn thành |
| **3** | **📌 Sinh dữ liệu & Load bằng Pandas** | **60 phút** | **Synthetic data · read_csv · dtype · parse_dates** | **▶ Đang học** |
| 4 | Làm sạch & Validate dữ liệu | 60 phút | Missing · Duplicate · Business rules · Quality report | ⬜ Chưa học |
| 5 | KPI & Cohort Retention Analysis | 60 phút | groupby · pivot_table · Cohort matrix · Heatmap | ⬜ Chưa học |
| 6 | Portfolio Reporting & Executive Summary | 60 phút | Visualization · Narrative · README · Executive summary | ⬜ Chưa học |
| 7 | Quiz & Checklist nộp bài | 45 phút | Final check · Self-assessment · Portfolio submission | ⬜ Chưa học |

---

> 💡 **Ghi nhớ:** Buổi này bạn sẽ **tự tạo ra dữ liệu** thay vì chờ dữ liệu từ công ty. Đây là kỹ năng thiết yếu để portfolio có thể reproduce được — bất kỳ ai cũng chạy lại được kết quả của bạn chỉ với một lệnh duy nhất.



# 🗃️ Buổi 3 — Sinh Dữ Liệu & Load Bằng Pandas

## 🎯 Mục tiêu buổi học

Kết thúc Buổi 3, bạn sẽ:

| # | Kết quả cụ thể |
|:-:|----------------|
| 1 | Hiểu khái niệm **synthetic data** và tại sao nó phù hợp cho portfolio |
| 2 | Đọc và hiểu được toàn bộ script `generate_ecommerce_dataset.py` từng dòng |
| 3 | Chạy script để sinh ra **4 file CSV** trong `data/raw/` |
| 4 | Load 4 bảng vào Pandas với đúng `dtype` và `parse_dates` |
| 5 | Thực hiện **schema validation**: kiểm tra shape, sample, unique key, referential integrity |

---

## 🔗 Kết nối với Buổi 2

```
Buổi 2 bạn đã làm:                    Buổi 3 bạn sẽ làm:
─────────────────────────────          ────────────────────────────────────
✅ Tạo folder structure               → Điền vào data/raw/ lần đầu tiên
✅ Thiết lập virtual environment      → Dùng numpy + pandas đã cài
✅ Viết business_brief.md             → Sinh dữ liệu đúng scope đã define
✅ Định nghĩa Metric Dictionary       → Schema CSV phải khớp metric definitions
✅ Tạo project_starter.ipynb skeleton → Hoàn thành Section 2 (Setup) & Section 3 (Load)
```

> ⚠️ **Quan trọng:** Tên cột trong 4 file CSV được thiết kế **khớp chính xác** với Metric Dictionary đã định nghĩa ở Buổi 2.  
> Ví dụ: metric **Revenue** định nghĩa là `SUM(net_revenue)` → cột `net_revenue` phải có trong `orders.csv`.  
> Metric **Active Customers** dùng `COUNT DISTINCT(customer_id)` với filter `status == 'completed'` → cột `customer_id` và `status` phải có trong `orders.csv`.

---

## 📦 Những gì chúng ta sẽ tạo ra hôm nay

```
cohort-retention-analytics/
└── data/
    └── raw/
        ├── customers.csv      ← 650 khách hàng
        ├── products.csv       ← 12 sản phẩm
        ├── orders.csv         ← ~1,500–2,500 đơn hàng (Jan–Dec 2025)
        └── order_items.csv    ← ~3,000–8,000 dòng chi tiết đơn hàng
```



## ⚠️ Môi Trường Thực Hành

> **Notebook này chỉ chạy được trên máy tính local của bạn.**  
> Tất cả code trong buổi này yêu cầu:  
> - Virtual environment đã được kích hoạt (đã làm ở Buổi 2)  
> - Thư mục project `cohort-retention-analytics/` đã tồn tại  
> - Các thư viện `numpy`, `pandas` đã được cài đặt  

---

### 🔁 Nhắc lại: Kích hoạt virtual environment trước khi chạy

**Windows (PowerShell):**
```powershell
cd cohort-retention-analytics
.\venv\Scripts\Activate.ps1
# Bạn thấy (venv) ở đầu dòng → đã vào môi trường ảo
```

**Mac / Linux (bash / zsh):**
```bash
cd cohort-retention-analytics
source venv/bin/activate
# Bạn thấy (venv) ở đầu dòng → đã vào môi trường ảo
```

**Kết quả mẫu**

![img3-1](images/images_buoi3/img3-1.png)

> 💡 **Mở `project_starter.ipynb`** (trong thư mục `notebooks/`) để chạy code từ Buổi 3 trở đi.  
> Notebook này (`buoi3_Data_Generation_Loading.ipynb`) là tài liệu học — đọc tại đây, **copy code vào `project_starter.ipynb`** để chạy.



## 🧪 Phần 1 — Dữ Liệu Tổng Hợp (Synthetic Data) Là Gì?

### 🤔 Vấn đề với dữ liệu thật trong portfolio

```
Dữ liệu thật của công ty:
  ❌ Không thể public lên GitHub → vi phạm NDA / bảo mật
  ❌ Người review không có dữ liệu → không thể re-run code của bạn
  ❌ Chứa PII (tên, email, phone) → rủi ro pháp lý
  ❌ Format, encoding, missing values không đồng nhất
```

```
Dữ liệu tổng hợp (synthetic data):
  ✅ Public hoàn toàn — không chứa thông tin cá nhân thật
  ✅ Reproducible — ai cũng có thể chạy lại từ đầu với cùng seed
  ✅ Được thiết kế để phù hợp đúng business case
  ✅ Clean và structured → focus vào analytics, không vào data wrangling
```

---

### 🌱 Seed là gì? Tại sao quan trọng?

Trong lập trình, số "ngẫu nhiên" thực ra là **giả ngẫu nhiên** (pseudo-random) — được tạo ra từ một công thức toán học bắt đầu bởi một giá trị gọi là **seed**.


In [ ]:
import numpy as np

# Seed giống nhau → kết quả giống nhau, dù chạy ở máy nào, lúc nào
rng = np.random.default_rng(seed=20260503)
print(rng.integers(1, 100, size=5))  # → [42, 17, 88, 63, 31] (luôn như vậy)

# Seed khác nhau → kết quả khác nhau
rng2 = np.random.default_rng(seed=999)
print(rng2.integers(1, 100, size=5)) # → [71, 55, 12, 97, 44] (khác)



[10 55 98  7 85]
[81 78 18 18 18]


> 💡 **Seed `20260503`** trong project này = ngày tháng (20/26/05/03 → gợi nhớ dễ).  
> Quan trọng nhất: **ai dùng seed này đều ra dataset giống hệt nhau** → kết quả phân tích reproducible.

---

### 📊 Dataset tổng quan

| Bảng | Grain (1 row = ?) | Số dòng | Thời gian |
|------|-------------------|:-------:|-----------|
| `customers.csv` | 1 khách hàng | ~650 | Signup: Jun 2024 – Dec 2025 |
| `products.csv` | 1 sản phẩm | 12 | — |
| `orders.csv` | 1 đơn hàng | ~1,500–2,500 | Order: Jan 2025 – Dec 2025 |
| `order_items.csv` | 1 sản phẩm trong 1 đơn | ~3,000–8,000 | — |
###
> 📌 **Tại sao có customers signup trước Jan 2025?**  
> Vì trong thực tế, khách hàng có thể signup từ trước khi bạn bắt đầu thu thập dữ liệu đơn hàng. Phân tích cohort cần `signup_date` để xác định cohort month — không phải `order_date` đầu tiên.


## 💻 Phần 2 — Script `generate_ecommerce_dataset.py`

### ⚡ Step 1 — Tạo file script

Tạo file `src/generate_ecommerce_dataset.py` trong thư mục project với nội dung sau.  
Đây là script Python thuần — không phải notebook. Copy toàn bộ nội dung bên dưới.

---

```python
#!/usr/bin/env python3
"""
generate_ecommerce_dataset.py

Sinh dữ liệu e-commerce tổng hợp (synthetic) cho dự án Cohort & Retention Analytics.
Tạo ra 4 file CSV: customers, products, orders, order_items.

Cách chạy:
    cd cohort-retention-analytics
    python src/generate_ecommerce_dataset.py --output data/raw --seed 20260503
"""

import argparse
from pathlib import Path

import numpy as np
import pandas as pd

# ─── Constants ────────────────────────────────────────────────────────────────
N_CUSTOMERS  = 650
N_PRODUCTS   = 12
ORDER_START  = "2025-01-01"   # Thời gian đặt hàng: Jan–Dec 2025
ORDER_END    = "2025-12-31"
SIGNUP_START = "2024-06-01"   # Một số khách signup trước kỳ phân tích


# ─── Generator: customers ─────────────────────────────────────────────────────
def gen_customers(rng: np.random.Generator) -> pd.DataFrame:
    """Sinh bảng customers (1 row = 1 khách hàng)."""
    n = N_CUSTOMERS

    # signup_date ngẫu nhiên trong khoảng Jun 2024 – Dec 2025
    signup_start = pd.Timestamp(SIGNUP_START)
    order_end    = pd.Timestamp(ORDER_END)
    total_days   = (order_end - signup_start).days
    offsets      = rng.integers(0, total_days + 1, size=n)
    signup_dates = [
        (signup_start + pd.Timedelta(days=int(d))).strftime("%Y-%m-%d")
        for d in offsets
    ]

    regions   = ["Hà Nội", "TP.HCM", "Đà Nẵng", "Hải Phòng", "Cần Thơ"]
    region_p  = [0.28, 0.42, 0.12, 0.10, 0.08]

    segments  = ["New", "Regular", "VIP"]
    segment_p = [0.52, 0.34, 0.14]

    channels  = ["organic", "paid_search", "social_media", "email", "referral"]
    channel_p = [0.26, 0.30, 0.20, 0.14, 0.10]

    return pd.DataFrame({
        "customer_id":         [f"CUST_{i+1:04d}" for i in range(n)],
        "signup_date":         signup_dates,
        "region":              rng.choice(regions,   size=n, p=region_p),
        "segment":             rng.choice(segments,  size=n, p=segment_p),
        "acquisition_channel": rng.choice(channels,  size=n, p=channel_p),
    })


# ─── Generator: products ──────────────────────────────────────────────────────
def gen_products() -> pd.DataFrame:
    """Sinh bảng products (1 row = 1 sản phẩm, cố định 12 sản phẩm)."""
    rows = [
        ("PROD_0001", "Electronics",  "TechVN",     2_500_000, 1_750_000),
        ("PROD_0002", "Electronics",  "TechVN",     4_800_000, 3_360_000),
        ("PROD_0003", "Electronics",  "SmartLife",    890_000,   622_000),
        ("PROD_0004", "Fashion",      "StyleVN",      350_000,   175_000),
        ("PROD_0005", "Fashion",      "StyleVN",      520_000,   260_000),
        ("PROD_0006", "Fashion",      "UrbanWear",    280_000,   140_000),
        ("PROD_0007", "Home & Living","HomeComfort",  620_000,   372_000),
        ("PROD_0008", "Home & Living","HomeComfort",1_200_000,   720_000),
        ("PROD_0009", "Beauty",       "GlowVN",       480_000,   240_000),
        ("PROD_0010", "Beauty",       "GlowVN",       320_000,   160_000),
        ("PROD_0011", "Sports",       "ActiveVN",     750_000,   450_000),
        ("PROD_0012", "Sports",       "ActiveVN",   1_100_000,   682_000),
    ]
    return pd.DataFrame(
        rows,
        columns=["product_id", "category", "brand", "list_price", "unit_cost"]
    )


# ─── Generator: orders + order_items ──────────────────────────────────────────
def gen_orders_and_items(
    rng: np.random.Generator,
    customers: pd.DataFrame,
    products: pd.DataFrame,
) -> tuple[pd.DataFrame, pd.DataFrame]:
    """Sinh bảng orders và order_items (có seasonality + segment behavior)."""

    # Số đơn trung bình theo segment (phân phối Poisson)
    avg_orders = {"New": 1.8, "Regular": 3.8, "VIP": 6.5}

    # Seasonality: tháng 7 (summer sale), tháng 11–12 (Black Friday + Tết)
    month_weights = {
         1: 0.070,  2: 0.060,  3: 0.070,
         4: 0.070,  5: 0.070,  6: 0.080,
         7: 0.105,  8: 0.080,  9: 0.070,
        10: 0.085, 11: 0.115, 12: 0.125,
    }
    months  = list(month_weights.keys())
    month_p = [month_weights[m] for m in months]

    # ~82% completed, ~10% cancelled, ~8% returned
    status_pool = ["completed"] * 82 + ["cancelled"] * 10 + ["returned"] * 8

    order_end    = pd.Timestamp(ORDER_END)
    orders_rows: list[dict] = []
    items_rows:  list[dict] = []
    oid_counter  = 1

    for _, cust in customers.iterrows():
        cid       = cust["customer_id"]
        seg       = cust["segment"]
        signup_ts = pd.Timestamp(cust["signup_date"])
        n_orders  = max(1, int(rng.poisson(avg_orders[seg])))

        for _ in range(n_orders):
            # Chọn tháng theo seasonality
            m       = rng.choice(months, p=month_p)
            m_start = pd.Timestamp(f"2025-{m:02d}-01")
            m_end   = m_start + pd.offsets.MonthEnd(0)
            days    = (m_end - m_start).days + 1
            order_date = m_start + pd.Timedelta(days=int(rng.integers(0, days)))

            # Đảm bảo order_date >= signup_date (không thể đặt hàng trước khi đăng ký)
            if order_date < signup_ts:
                order_date = signup_ts + pd.Timedelta(days=int(rng.integers(1, 15)))
            order_date = min(order_date, order_end)

            status = rng.choice(status_pool)
            oid    = f"ORD_{oid_counter:05d}"
            oid_counter += 1

            # Sinh order_items: mỗi đơn có 1–6 sản phẩm khác nhau
            n_items  = int(rng.choice([1, 2, 3, 4, 5, 6], p=[0.30, 0.28, 0.20, 0.12, 0.06, 0.04]))
            prod_idx = rng.choice(len(products), size=n_items, replace=False)
            gross    = 0.0

            for pi in prod_idx:
                prod = products.iloc[pi]
                # Phân phối quantity thực tế: long-tail 1–100
                # ~70% đơn lẻ (1–5), ~20% mua nhiều (6–20), ~10% mua sỉ (21–100)
                tier = rng.random()
                if tier < 0.70:
                    qty = int(rng.integers(1, 6))       # 1–5
                elif tier < 0.90:
                    qty = int(rng.integers(6, 21))      # 6–20
                else:
                    qty = int(rng.integers(21, 101))    # 21–100
                # unit_price = list_price ± 5%, làm tròn đến 1,000 VNĐ
                up = float(prod["list_price"]) * rng.uniform(0.95, 1.05)
                up = round(up / 1000) * 1000
                items_rows.append({
                    "order_id":   oid,
                    "product_id": prod["product_id"],
                    "quantity":   qty,
                    "unit_price": int(up),
                })
                gross += up * qty

            # net_revenue: completed=gross, returned=10% (phí xử lý), cancelled=0
            if status == "completed":
                net = gross
            elif status == "returned":
                net = round(gross * 0.10, 2)
            else:  # cancelled
                net = 0.0

            orders_rows.append({
                "order_id":      oid,
                "customer_id":   cid,
                "order_date":    order_date.strftime("%Y-%m-%d"),
                "status":        status,
                "gross_revenue": round(gross, 2),
                "net_revenue":   round(net, 2),
            })

    return pd.DataFrame(orders_rows), pd.DataFrame(items_rows)


# ─── Main ─────────────────────────────────────────────────────────────────────
def main() -> None:
    parser = argparse.ArgumentParser(
        description="Sinh e-commerce dataset tổng hợp cho Cohort & Retention Analytics"
    )
    parser.add_argument(
        "--output", default="data/raw",
        help="Thư mục lưu file CSV (default: data/raw)"
    )
    parser.add_argument(
        "--seed", type=int, default=20260503,
        help="Random seed để đảm bảo reproducibility (default: 20260503)"
    )
    args = parser.parse_args()

    out_dir = Path(args.output)
    out_dir.mkdir(parents=True, exist_ok=True)

    rng = np.random.default_rng(args.seed)

    print(f"🌱 Random seed  : {args.seed}")
    print(f"📁 Output folder: {out_dir.resolve()}")
    print("⏳ Generating dataset...\n")

    customers              = gen_customers(rng)
    products               = gen_products()
    orders, order_items    = gen_orders_and_items(rng, customers, products)

    customers.to_csv(   out_dir / "customers.csv",   index=False, encoding="utf-8-sig")
    products.to_csv(    out_dir / "products.csv",    index=False, encoding="utf-8-sig")
    orders.to_csv(      out_dir / "orders.csv",      index=False, encoding="utf-8-sig")
    order_items.to_csv( out_dir / "order_items.csv", index=False, encoding="utf-8-sig")

    print(f"✅ customers.csv     → {len(customers):>6,} rows")
    print(f"✅ products.csv      → {len(products):>6,} rows")
    print(f"✅ orders.csv        → {len(orders):>6,} rows")
    print(f"✅ order_items.csv   → {len(order_items):>6,} rows")
    print("\n🎉 Dataset generated successfully!")


if __name__ == "__main__":
    main()
```



### 🔍 Giải thích script từng phần

| Phần | Code | Ý nghĩa |
|------|------|---------|
| **Constants** | `N_CUSTOMERS`, `ORDER_START`, `SIGNUP_START` | Định nghĩa tập trung → thay đổi 1 chỗ, ảnh hưởng toàn bộ script |
| **`gen_customers()`** | `rng.integers(0, total_days+1, size=n)` | Sinh `n` offsets ngẫu nhiên → cộng vào `signup_start` → ra `signup_date` |
| **`gen_customers()`** | `rng.choice(regions, size=n, p=region_p)` | Chọn ngẫu nhiên có trọng số → TP.HCM chiếm 42%, Cần Thơ chỉ 8% |
| **`gen_products()`** | Hard-coded 12 sản phẩm | Sản phẩm cố định → giá ổn định, dễ kiểm tra |
| **`gen_orders_and_items()`** | `rng.poisson(avg_orders[seg])` | Phân phối Poisson → mô phỏng số đơn thực tế (có outlier tự nhiên) |
| **Seasonality** | `month_weights` | Tháng 11–12 có trọng số cao hơn → mô phỏng Black Friday + Tết |
| **Status logic** | `status_pool` | 82% completed / 10% cancelled / 8% returned |
| **net_revenue** | `if status == "completed": net = gross` | Khớp với Metric Dictionary: Revenue chỉ tính đơn `completed` |
| **`--seed`** | `np.random.default_rng(args.seed)` | `default_rng` là generator hiện đại, an toàn hơn `np.random.seed()` |

---

### 📌 Tại sao dùng `np.random.default_rng()` thay vì `random.seed()`?

In [9]:
# ❌ Cách cũ — global state, không thread-safe
import random
random.seed(42)
random.randint(1, 100)

82

In [10]:
# ❌ Cách cũ của numpy — bị deprecated
import numpy as np
np.random.seed(42)
np.random.randint(1, 100)

52

In [11]:
# ✅ Cách chuẩn hiện nay — tạo Generator object độc lập, thread-safe
rng = np.random.default_rng(seed=42)
rng.integers(1, 101)   # lưu ý: integers(low, high) trong default_rng là [low, high)

np.int64(9)

> 💡 `np.random.default_rng()` là chuẩn từ **NumPy 1.17+**. Trong production code và data science projects, luôn dùng cách này.


## ⚡ Phần 3 — Chạy Script Sinh Dữ Liệu

### Step 2 — Tạo thư mục `src/` và lưu script

**Windows (PowerShell):**
```powershell
cd cohort-retention-analytics
New-Item -ItemType Directory -Force -Path src
# Sau đó tạo file src/generate_ecommerce_dataset.py với nội dung ở trên
```

**Mac / Linux:**
```bash
cd cohort-retention-analytics
mkdir -p src
# Sau đó tạo file src/generate_ecommerce_dataset.py với nội dung ở trên
```
**Kết quả mẫu khởi tại file generate_ecommerce_dataset.py**

![img3-2](images/images_buoi3/img3-2.png)

---

### Step 3 — Chạy script để sinh dữ liệu

Đảm bảo bạn đang ở trong thư mục `cohort-retention-analytics/` và virtual environment đã được kích hoạt:

**Windows (PowerShell):**
```powershell
python src/generate_ecommerce_dataset.py --output data/raw --seed 20260503
```

**Mac / Linux:**
```bash
python src/generate_ecommerce_dataset.py --output data/raw --seed 20260503
```

---

### 📋 Kết quả mong đợi trên console

```Chạy file generate_ecommerce_dataset.py----->```

![img3-3](images/images_buoi3/img3-3.png)

> 📌 Số dòng trong `orders.csv` và `order_items.csv` có thể thay đổi nhẹ tùy seed, nhưng sẽ **luôn trong khoảng 1,500–2,500** và **3,000–8,000** tương ứng.

---

### Step 4 — Xác nhận file đã được tạo

**Windows (PowerShell):**
```powershell
Get-ChildItem data/raw | Select-Object Name, Length
```

**Mac / Linux:**
```bash
ls -lh data/raw/
```

**Kết quả mong đợi:**

![img3-4](images/images_buoi3/img3-4.png)

> ✅ Nếu bạn thấy 4 file trên → dữ liệu đã sẵn sàng để load vào Pandas!



## 🗂️ Phần 4 — Schema Tham Chiếu (Data Dictionary)

Hiểu rõ schema trước khi load — đây là bước **data contract**: mô tả chính xác những gì mỗi bảng chứa.

---

### 📋 Bảng 1: `customers.csv` — Grain: 1 row = 1 khách hàng

| Cột | Kiểu dữ liệu | Mô tả | Ví dụ |
|-----|-------------|-------|-------|
| `customer_id` | string | **Primary Key** — mã khách hàng duy nhất | `CUST_0001` |
| `signup_date` | date | Ngày khách hàng đăng ký tài khoản | `2024-08-15` |
| `region` | string (categorical) | Khu vực địa lý | `TP.HCM` |
| `segment` | string (categorical) | Phân khúc khách hàng | `Regular` |
| `acquisition_channel` | string (categorical) | Kênh tiếp thị đã acquire | `paid_search` |

**Giá trị hợp lệ:**
- `region`: `Hà Nội`, `TP.HCM`, `Đà Nẵng`, `Hải Phòng`, `Cần Thơ`
- `segment`: `New`, `Regular`, `VIP`
- `acquisition_channel`: `organic`, `paid_search`, `social_media`, `email`, `referral`

---

### 📋 Bảng 2: `products.csv` — Grain: 1 row = 1 sản phẩm

| Cột | Kiểu dữ liệu | Mô tả | Ví dụ |
|-----|-------------|-------|-------|
| `product_id` | string | **Primary Key** — mã sản phẩm | `PROD_0001` |
| `category` | string (categorical) | Danh mục sản phẩm | `Electronics` |
| `brand` | string | Thương hiệu | `TechVN` |
| `list_price` | integer (VNĐ) | Giá niêm yết | `2,500,000` |
| `unit_cost` | integer (VNĐ) | Giá vốn | `1,750,000` |

**Giá trị hợp lệ `category`:** `Electronics`, `Fashion`, `Home & Living`, `Beauty`, `Sports`

---

### 📋 Bảng 3: `orders.csv` — Grain: 1 row = 1 đơn hàng

| Cột | Kiểu dữ liệu | Mô tả | Ví dụ |
|-----|-------------|-------|-------|
| `order_id` | string | **Primary Key** — mã đơn hàng | `ORD_00001` |
| `customer_id` | string | **Foreign Key** → `customers.customer_id` | `CUST_0042` |
| `order_date` | date | Ngày đặt hàng | `2025-07-14` |
| `status` | string (categorical) | Trạng thái đơn hàng | `completed` |
| `gross_revenue` | float (VNĐ) | Tổng doanh thu trước điều chỉnh | `1,560,000` |
| `net_revenue` | float (VNĐ) | Doanh thu thực (sau cancelled/returned) | `1,560,000` |

**Giá trị hợp lệ `status`:** `completed`, `cancelled`, `returned`  
**Logic `net_revenue`:**
- `completed` → `net_revenue = gross_revenue`
- `returned`  → `net_revenue = gross_revenue × 10%` (phí xử lý)
- `cancelled`  → `net_revenue = 0`

---

### 📋 Bảng 4: `order_items.csv` — Grain: 1 row = 1 sản phẩm trong 1 đơn

| Cột | Kiểu dữ liệu | Mô tả | Ví dụ |
|-----|-------------|-------|-------|
| `order_id` | string | **Foreign Key** → `orders.order_id` | `ORD_00001` |
| `product_id` | string | **Foreign Key** → `products.product_id` | `PROD_0003` |
| `quantity` | integer | Số lượng sản phẩm | `2` |
| `unit_price` | integer (VNĐ) | Giá thực tế khi mua | `880,000` |

---

### 🔗 Kết nối với Metric Dictionary (Buổi 2)

```
Metric                Formula (Buổi 2)                 Cột tương ứng (Buổi 3)
──────────────────────────────────────────────────────────────────────────────
Revenue             → SUM(net_revenue)              →  orders.net_revenue
                       filter: status='completed'   →  orders.status
Orders              → COUNT DISTINCT(order_id)      →  orders.order_id
Active Customers    → COUNT DISTINCT(customer_id)   →  orders.customer_id
AOV                 → Revenue / Orders              →  tính từ 2 cột trên
Cohort Size         → COUNT DISTINCT(customer_id)   →  customers.customer_id
                       tại cohort_index=0           →  + customers.signup_date
Retention Rate      → Active / Cohort Size          →  kết hợp cả 2 bảng
```

> ✅ **Schema đã được thiết kế 100% khớp với Metric Dictionary.** Không cần rename, không cần transform tên cột khi tính toán từ Buổi 5.



## 💻 Phần 5 — Load Dữ Liệu Vào Pandas

### ⚡ Step 5 — Hoàn thành Section 2 của `project_starter.ipynb`

> 📌 Copy toàn bộ code trong **Cell 8 (Python)** dưới đây vào **Section 2: Setup & Imports** của `project_starter.ipynb`.  
> Lưu ý: Biến `PROJECT_DIR`, `RAW_DIR`, `PROCESSED_DIR`, `OUTPUT_TABLE_DIR`, `OUTPUT_CHART_DIR` là **tên chuẩn** sẽ dùng xuyên suốt từ Buổi 3 đến Buổi 6.

---

### 🔄 Cập nhật từ skeleton Buổi 2

Trong Buổi 2, skeleton `project_starter.ipynb` dùng tên biến rút gọn:
```python
# Buổi 2 skeleton (cũ) — sẽ được thay thế
ROOT       = Path.cwd().parent
DATA_RAW   = ROOT / 'data' / 'raw'
DATA_PROC  = ROOT / 'data' / 'processed'
OUT_TABLES = ROOT / 'outputs' / 'tables'
OUT_CHARTS = ROOT / 'outputs' / 'charts'
```

Từ Buổi 3, chúng ta dùng tên đầy đủ hơn và đường dẫn chính xác:
```python
# Buổi 3+ (chuẩn) — dùng tên này từ đây trở đi
# Notebook nằm trong "Cohort & Retention Analytics Portfolio" (folder học tập)
# Data & outputs nằm trong "cohort-retention-analytics" (folder project — cùng cấp)
PROJECT_DIR      = Path.cwd().parent / "cohort-retention-analytics"
RAW_DIR          = PROJECT_DIR / "data" / "raw"
PROCESSED_DIR    = PROJECT_DIR / "data" / "processed"
OUTPUT_TABLE_DIR = PROJECT_DIR / "outputs" / "tables"
OUTPUT_CHART_DIR = PROJECT_DIR / "outputs" / "charts"
```

> **Thay thế toàn bộ** phần path trong `project_starter.ipynb` bằng chuẩn mới này.


In [ ]:

# ═══════════════════════════════════════════════════════════════════════════
# Section 2: Setup & Imports
# ═══════════════════════════════════════════════════════════════════════════

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from pathlib import Path

# ── Display options ────────────────────────────────────────────────────────
pd.set_option("display.max_columns", None)
pd.set_option("display.float_format", "{:,.0f}".format)
plt.style.use("seaborn-v0_8-whitegrid")

# ── Project paths ──────────────────────────────────────────────────────────
# Notebook này nằm trong notebooks/ — một subfolder của project root.
# Path.cwd()        = C:\Users\ADMIN\cohort-retention-analytics\notebooks
# Path.cwd().parent = C:\Users\ADMIN\cohort-retention-analytics  ← project root
PROJECT_DIR      = Path.cwd().parent
RAW_DIR          = PROJECT_DIR / "data" / "raw"
PROCESSED_DIR    = PROJECT_DIR / "data" / "processed"
OUTPUT_TABLE_DIR = PROJECT_DIR / "outputs" / "tables"
OUTPUT_CHART_DIR = PROJECT_DIR / "outputs" / "charts"

for d in [PROCESSED_DIR, OUTPUT_TABLE_DIR, OUTPUT_CHART_DIR]:
    d.mkdir(parents=True, exist_ok=True)

print("✅ Setup complete")
print(f"   PROJECT_DIR      : {PROJECT_DIR}")
print(f"   RAW_DIR          : {RAW_DIR}")
print(f"   PROCESSED_DIR    : {PROCESSED_DIR}")
print(f"   OUTPUT_TABLE_DIR : {OUTPUT_TABLE_DIR}")
print(f"   OUTPUT_CHART_DIR : {OUTPUT_CHART_DIR}")
print(f"\n   RAW_DIR exists   : {RAW_DIR.exists()}")


**Kết quả mẫu khi chạy cell 4 bên project_starter.ipynb**

![img3-5](images/images_buoi3/img3-5.png)


### 🔍 Giải thích từng dòng code Setup

```python
pd.set_option("display.max_columns", None)
```
> Mặc định Pandas ẩn bớt cột nếu DataFrame có nhiều cột. `None` = hiện tất cả.

```python
pd.set_option("display.float_format", "{:,.0f}".format)
```
> Format số thực: `1560000.0` → `1,560,000`. Dễ đọc hơn khi xem doanh thu VNĐ.

```python
plt.style.use("seaborn-v0_8-whitegrid")
```
> Style biểu đồ nhất quán. `seaborn-v0_8-*` là tên đúng từ matplotlib 3.6+.

```python
PROJECT_DIR = Path.cwd().parent / "cohort-retention-analytics"
```
> **Tại sao cần chỉ rõ như vậy?**  
> Project có **2 folder riêng biệt** cùng nằm trong `C:\Users\ADMIN\`:  
> - `Cohort & Retention Analytics Portfolio\` → notebook học tập (file này đang chạy)  
> - `cohort-retention-analytics\` → project thật: data, outputs, scripts  
>
> `Path.cwd()` khi chạy notebook = `C:\Users\ADMIN\Cohort & Retention Analytics Portfolio\`  
> `Path.cwd().parent` = `C:\Users\ADMIN\`  
> `Path.cwd().parent / "cohort-retention-analytics"` = `C:\Users\ADMIN\cohort-retention-analytics\` ✅  
>
> ⚠️ **Sai lầm phổ biến**: dùng `Path.cwd()` thẳng → trỏ vào folder notebook → `mkdir` tạo thư mục `data/` và `outputs/` **trong folder học tập** (sai hoàn toàn).

```python
for d in [PROCESSED_DIR, OUTPUT_TABLE_DIR, OUTPUT_CHART_DIR]:
    d.mkdir(parents=True, exist_ok=True)
```
> `parents=True` → tạo cả thư mục cha nếu chưa tồn tại.  
> `exist_ok=True` → không raise error nếu thư mục đã có.  
> `RAW_DIR` không cần tạo vì script đã tạo khi sinh dữ liệu.



### ⚡ Step 6 — Load 4 bảng với `pd.read_csv()`

> 📌 Copy code dưới đây vào **Section 3: Load Data** của `project_starter.ipynb`.

**Vì sao cần `parse_dates`?**

```python
# ❌ Không dùng parse_dates → cột date bị đọc là string
df = pd.read_csv("orders.csv")
df["order_date"].dtype  # → object (string)
df["order_date"] > "2025-06-01"  # so sánh string, có thể sai!

# ✅ Dùng parse_dates → cột date được nhận dạng đúng
df = pd.read_csv("orders.csv", parse_dates=["order_date"])
df["order_date"].dtype  # → datetime64[ns]
df["order_date"] > pd.Timestamp("2025-06-01")  # so sánh datetime, đúng
df["order_date"].dt.to_period("M")             # lấy tháng → dùng cho cohort
```


In [ ]:

# ═══════════════════════════════════════════════════════════════════════════
# Section 3: Load Data
# → Thêm vào project_starter.ipynb — chạy sau Section 2
# ═══════════════════════════════════════════════════════════════════════════
# ── Load 4 bảng từ data/raw/ ───────────────────────────────────────────────
customers = pd.read_csv(
    RAW_DIR / "customers.csv",
    parse_dates=["signup_date"],    # đọc signup_date là datetime64, không phải string
)

products = pd.read_csv(
    RAW_DIR / "products.csv",
    # list_price và unit_cost là integer → Pandas tự nhận dạng đúng
)

orders = pd.read_csv(
    RAW_DIR / "orders.csv",
    parse_dates=["order_date"],     # đọc order_date là datetime64
)

order_items = pd.read_csv(
    RAW_DIR / "order_items.csv",
    # tất cả cột đều là string/integer → không cần parse_dates
)

# ── Kiểm tra nhanh sau khi load ─────────────────────────────────────────────
print("✅ Loaded 4 tables:")
for name, df in [
    ("customers",   customers),
    ("products",    products),
    ("orders",      orders),
    ("order_items", order_items),
]:
    print(f"   {name:<15} → {df.shape[0]:>6,} rows  ×  {df.shape[1]} cols")


**Kết quả mẫu khi chạy cell 6 bên project_starter.ipynb**

![img3-6](images/images_buoi3/img3-6.png)


## 🔬 Phần 6 — Kiểm Tra Cấu Trúc Dữ Liệu

Sau khi load, luôn thực hiện 3 kiểm tra cơ bản:

| Kiểm tra | Câu hỏi | Method |
|----------|---------|--------|
| **Shape & sample** | Bao nhiêu dòng, bao nhiêu cột? Dữ liệu trông như thế nào? | `df.shape`, `df.head()`, `df.dtypes` |
| **Uniqueness** | Key column có thực sự unique không? | `df[col].nunique() == len(df)` |
| **Referential integrity** | Foreign key có trỏ đến bảng đúng không? | `df[fk].isin(ref[pk]).all()` |

---

### ⚡ Step 7 — Shape, dtypes và sample


In [ ]:

# ═══════════════════════════════════════════════════════════════════════════
# Section 3 (tiếp): Shape, dtypes, sample — kiểm tra tất cả 4 bảng
# ═══════════════════════════════════════════════════════════════════════════

tables = {
    "customers":   customers,
    "products":    products,
    "orders":      orders,
    "order_items": order_items,
}

for tbl_name, df in tables.items():
    print(f"\n{'═'*60}")
    print(f"  📋  {tbl_name.upper()}  —  shape: {df.shape[0]:,} rows × {df.shape[1]} cols")
    print(f"{'═'*60}")
    print(df.dtypes.to_string())
    print(f"\n--- Sample (head 3) ---")
    display(df.head(3))


**Kết quả mẫu khi chạy cell 7 bên project_starter.ipynb**

![img3-7](images/images_buoi3/img3-7.png)

```----->```

![img3-8](images/images_buoi3/img3-8.png)

```----->```

![img3-9](images/images_buoi3/img3-9.png)

```----->```

![img3-10](images/images_buoi3/img3-10.png)


### 🔍 Đọc kết quả dtypes

Sau khi chạy cell trên, bạn sẽ thấy dtypes như sau:

**`customers`:**
```
customer_id              object      ← string (đúng — không cần numeric)
signup_date      datetime64[ns]      ← datetime (đúng — parse_dates hoạt động)
region                   object      ← string categorical (đúng)
segment                  object      ← string categorical (đúng)
acquisition_channel      object      ← string categorical (đúng)
```

**`orders`:**
```
order_id                 object      ← string (đúng)
customer_id              object      ← string Foreign Key (đúng)
order_date       datetime64[ns]      ← datetime (đúng — parse_dates hoạt động)
status                   object      ← string categorical (đúng)
gross_revenue           float64      ← số thực VNĐ (đúng)
net_revenue             float64      ← số thực VNĐ (đúng)
```

> ⚠️ **Cờ đỏ:** Nếu bạn thấy `signup_date` hoặc `order_date` là `object` thay vì `datetime64` → `parse_dates` không hoạt động → kiểm tra lại file CSV có cột đó không.

---

### 🔑 Tại sao không convert `customer_id` sang integer?

```python
# Câu hỏi phổ biến: "CUST_0001" → tại sao không lưu số 1 thôi?"

# Trả lời: ID là định danh, không phải số để tính toán
# CUST_0001 + CUST_0002 không có nghĩa gì
# Dùng string format CUST_XXXX:
#   ✅ Rõ ràng đây là customer ID, không phải số lượng
#   ✅ Dễ join với hệ thống khác (CRM, BI tool)
#   ✅ Tránh nhầm lẫn khi merge nhiều bảng
```

---

### ⚡ Step 8 — Key uniqueness & Referential Integrity check


In [ ]:
# ═══════════════════════════════════════════════════════════════════════════
# Section 3 (tiếp): Key uniqueness & Referential Integrity
# Kiểm tra: primary key unique? foreign key hợp lệ?
# ═══════════════════════════════════════════════════════════════════════════

checks = {
    "customers  : customer_id  là PK duy nhất":  customers["customer_id"].nunique()  == len(customers),
    "products   : product_id   là PK duy nhất":  products["product_id"].nunique()    == len(products),
    "orders     : order_id     là PK duy nhất":  orders["order_id"].nunique()        == len(orders),
    "orders     : customer_id  → customers":     orders["customer_id"].isin(customers["customer_id"]).all(),
    "order_items: order_id     → orders":        order_items["order_id"].isin(orders["order_id"]).all(),
    "order_items: product_id   → products":      order_items["product_id"].isin(products["product_id"]).all(),
    "orders     : status chỉ 3 giá trị hợp lệ": orders["status"].isin(["completed", "cancelled", "returned"]).all(),
    "orders     : net_revenue  >= 0":            (orders["net_revenue"] >= 0).all(),
    "order_items: quantity     >= 1":            (order_items["quantity"] >= 1).all(),
    "order_items: unit_price   > 0":             (order_items["unit_price"] > 0).all(),
}

result_df = pd.DataFrame({
    "Kiểm tra": checks.keys(),
    "Kết quả":  ["✅ PASS" if v else "❌ FAIL" for v in checks.values()],
})
display(result_df.style.hide(axis="index"))

all_pass = all(checks.values())
print(f"\n{'✅ Tất cả checks đều PASS — dữ liệu hợp lệ!' if all_pass else '❌ Có check thất bại — kiểm tra lại dữ liệu!'}")


**Kết quả mẫu khi chạy cell 8 bên project_starter.ipynb**

![img3-11](images/images_buoi3/img3-11.png)


### 🔍 Giải thích các loại kiểm tra

**📌 Primary Key Uniqueness:**
```python
customers["customer_id"].nunique() == len(customers)
# nunique() = số giá trị distinct trong cột
# len(df)   = tổng số dòng
# Nếu bằng nhau → không có duplicate → PK hợp lệ
```

**📌 Referential Integrity:**
```python
orders["customer_id"].isin(customers["customer_id"]).all()
# isin() → trả về Series[bool]: True nếu giá trị tồn tại trong tập tham chiếu
# .all() → True nếu tất cả đều True
# Kết quả: mọi customer_id trong orders đều có trong customers → FK hợp lệ
```

**📌 Business Rule Validation:**
```python
orders["status"].isin(["completed", "cancelled", "returned"]).all()
# Kiểm tra không có giá trị lạ ngoài 3 trạng thái đã định nghĩa
```

> 💡 **Thực tế:** Đây là bước **Data Quality Gate** đầu tiên. Nếu bất kỳ check nào FAIL, bạn phải tìm nguyên nhân **trước khi** tiến hành phân tích — số liệu sẽ sai nếu FK có orphan records.

---

### ⚡ Step 9 — Khám phá nhanh phân phối dữ liệu


In [ ]:
# ═══════════════════════════════════════════════════════════════════════════
# Section 3 (tiếp): Khám phá nhanh phân phối
# Mục đích: sanity check — dữ liệu có vẻ hợp lý không?
# ═══════════════════════════════════════════════════════════════════════════

print("── Customers: phân phối theo segment ──────────────────────────")
print(customers["segment"].value_counts().to_string())

print("\n── Customers: phân phối theo acquisition_channel ──────────────")
print(customers["acquisition_channel"].value_counts().to_string())

print("\n── Orders: phân phối theo status ───────────────────────────────")
status_counts = orders["status"].value_counts()
status_pct    = (orders["status"].value_counts(normalize=True) * 100).round(1)
status_summary = pd.DataFrame({"count": status_counts, "pct (%)": status_pct})
print(status_summary.to_string())

print("\n── Orders: phân phối theo tháng (seasonality check) ───────────")
monthly_counts = (
    orders["order_date"]
    .dt.to_period("M")
    .value_counts()
    .sort_index()
)
print(monthly_counts.to_string())

print("\n── Orders: thống kê net_revenue (chỉ completed) ────────────────")
completed = orders[orders["status"] == "completed"]["net_revenue"]
print(f"  Count   : {len(completed):,}")
print(f"  Sum     : {completed.sum():,.0f} VNĐ")
print(f"  Mean    : {completed.mean():,.0f} VNĐ")
print(f"  Median  : {completed.median():,.0f} VNĐ")
print(f"  Min     : {completed.min():,.0f} VNĐ")
print(f"  Max     : {completed.max():,.0f} VNĐ")

**Kết quả mẫu khi chạy cell 8 bên project_starter.ipynb**

![img3-12](images/images_buoi3/img3-12.png)


### 🔍 Những gì cần kiểm tra từ kết quả exploration

**Phân phối segment hợp lý không?**
> Mong đợi: `New ~52%`, `Regular ~34%`, `VIP ~14%`  
> Nếu tỷ lệ quá lệch (ví dụ 90%/5%/5%) → seed bị sai hoặc script có bug

**Seasonality có đúng không?**
> Mong đợi: tháng 11, 12 có nhiều đơn nhất; tháng 2 ít nhất  
> Đây là behaviour đã thiết kế trong `month_weights` của script

**Phân phối status hợp lý không?**
> Mong đợi: `completed ~82%`, `cancelled ~10%`, `returned ~8%`  
> Nếu bạn thấy tỷ lệ khác xa → kiểm tra lại `status_pool` trong script

**Net revenue có sanity check không?**
> `completed` orders: `net_revenue = gross_revenue > 0` ✅  
> Tổng net revenue (chỉ completed) = doanh thu thực tế của 12 tháng → con số này sẽ xuất hiện lại trong Buổi 5 (Monthly KPI)

---

## 📊 Phần 7 — Kết Nối Với Buổi Tiếp Theo

Sau Buổi 3, trong `project_starter.ipynb` bạn đã hoàn thành:

```
Section 1: Business Questions    ✅ (từ Buổi 2 — business_brief.md)
Section 2: Setup & Imports        ✅ (Buổi 3 — PROJECT_DIR, RAW_DIR, thư viện)
Section 3: Load Data              ✅ (Buổi 3 — 4 DataFrame đã load & validated)
Section 4: Data Quality Checks    ← Buổi 4 sẽ làm
Section 5: Cleaning Decisions     ← Buổi 4 sẽ làm
Section 6: KPI Analysis           ← Buổi 5 sẽ làm
Section 7: Cohort Retention       ← Buổi 5 sẽ làm
Section 8: Visualization          ← Buổi 6 sẽ làm
Section 9: Executive Summary      ← Buổi 6 sẽ làm
Section 10: Export Outputs        ← Buổi 6 sẽ làm
```

**Buổi 4 sẽ dùng:**
- `customers`, `products`, `orders`, `order_items` (đã load ở Buổi 3)  
- `PROJECT_DIR`, `RAW_DIR`, `PROCESSED_DIR` (path variables đã định nghĩa)  
- Làm sạch data: missing values, outliers, business rule violations  
- Xuất cleaned data ra `data/processed/`



## ✍️ Bài Tập Buổi 3

### 📝 Bài tập 1 — Chạy và xác nhận script

Chạy script sinh dữ liệu và xác nhận:
- [ ] 4 file CSV đã tồn tại trong `data/raw/`
- [ ] `customers.csv` có đúng 650 dòng
- [ ] `products.csv` có đúng 12 dòng
- [ ] `orders.csv` có từ 1,500 đến 2,500 dòng
- [ ] `order_items.csv` có từ 3,000 đến 8,000 dòng

---

### 📝 Bài tập 2 — Load và validate

Trong `project_starter.ipynb`, điền đầy đủ Section 2 và Section 3, sau đó kiểm tra:

**2a.** Chạy cell setup — `PROJECT_DIR` có trỏ đến đúng `cohort-retention-analytics/` không?

**2b.** Sau khi load, kiểm tra `orders["order_date"].dtype`. Kết quả là gì?  
*(Đáp án đúng: `datetime64[ns]`)*

**2c.** Đếm số đơn hàng theo từng status:
```python
orders["status"].value_counts()
```
Tỷ lệ `completed` của bạn là bao nhiêu phần trăm?

---

### 📝 Bài tập 3 — Khám phá dữ liệu

**3a.** Tìm `customer_id` có nhiều đơn hàng nhất:
```python
orders.groupby("customer_id")["order_id"].count().sort_values(ascending=False).head(5)
```

**3b.** Tổng `net_revenue` của toàn bộ đơn `completed` trong năm 2025 là bao nhiêu?
```python
orders[orders["status"] == "completed"]["net_revenue"].sum()
```

**3c.** Category nào có `list_price` trung bình cao nhất?
```python
products.groupby("category")["list_price"].mean().sort_values(ascending=False)
```

---

### ✅ Checklist Buổi 3 — Hoàn thành trước Buổi 4

- [ ] Script `src/generate_ecommerce_dataset.py` đã được tạo
- [ ] 4 file CSV có trong `data/raw/` với đúng số dòng
- [ ] `project_starter.ipynb` Section 2 (Setup) đã hoàn chỉnh với tên biến chuẩn
- [ ] `project_starter.ipynb` Section 3 (Load Data) đã hoàn chỉnh
- [ ] Tất cả 10 integrity checks đều PASS (✅)
- [ ] `signup_date` và `order_date` đều có dtype `datetime64[ns]`
- [ ] Tên 4 DataFrame: `customers`, `products`, `orders`, `order_items`
- [ ] Path variables: `PROJECT_DIR`, `RAW_DIR`, `PROCESSED_DIR`, `OUTPUT_TABLE_DIR`, `OUTPUT_CHART_DIR`
- [ ] Bạn hiểu tại sao `status_pool` trong script thiết kế ~82% completed

---

### 🔜 Chuẩn bị Buổi 4 — Data Cleaning & Validation

Buổi 4 sẽ:
1. Kiểm tra **missing values** — cột nào có null? Tại sao? Xử lý như thế nào?
2. Kiểm tra **duplicate rows** — phân biệt exact duplicate vs logical duplicate
3. Kiểm tra **outliers** trong `net_revenue` — IQR method, business rule validation
4. Xây dựng **`orders_clean`** — bảng đã làm sạch, lưu ra `data/processed/orders_clean.csv`
5. Ghi lại **Cleaning Decision Log** — mọi quyết định làm sạch đều phải có lý do

> 💡 **Lưu ý:** Trong Buổi 4, tất cả code đều dùng `orders`, `customers`, `products`, `order_items` và `PROCESSED_DIR` đã định nghĩa ở Buổi 3 — không cần re-load, chỉ cần run notebook từ đầu.
